src/streaming_kzg/mod.rs

```diff
@@ -0,0 +1,325 @@
+        proof: &EvaluationProof<E>,
+        open_chal: &E::Fr,
+    ) -> VerificationResult {
+        // Computing the vanishing polynomial over eval_points
+        let zeros = vanishing_polynomial(eval_points);
+        let zeros_repr = zeros.iter().map(|x| x.into_bigint()).collect::<Vec<_>>();
+        let zeros = VariableBase::msm(&self.powers_of_g2, &zeros_repr);
+
+        // Computing the inverse for the interpolation
+        let mut sca_inverse = Vec::new();
+        for (j, x_j) in eval_points.iter().enumerate() {
+            let mut sca = E::Fr::one();
+            for (k, x_k) in eval_points.iter().enumerate() {
+                if j == k {
+                    continue;
+                }
+                sca *= *x_j - x_k;
+            }
+            sca = sca.inverse().unwrap();
+            sca_inverse.push(sca);
+        }
+
+        // Computing the lagrange polynomial for the interpolation
+        let mut lang = Vec::new();
+        for (j, _x_j) in eval_points.iter().enumerate() {
+            let mut l_poly = DensePolynomial::from_coefficients_vec(vec![E::Fr::one()]);
+            for (k, x_k) in eval_points.iter().enumerate() {
+                if j == k {
+                    continue;
+                }
+                let tmp_poly = DensePolynomial::from_coefficients_vec(vec![-(*x_k), E::Fr::one()]);
+                l_poly = l_poly.mul(&tmp_poly);
+            }
+            lang.push(l_poly);
+        }
+
+        // Computing the commitment for the interpolated polynomials
+        let etas = powers(*open_chal, evaluations.len());
+        let interpolated_polynomials = evaluations
+            .iter()
+            .map(|e| interpolate_poly::<E>(eval_points, e, &sca_inverse, &lang).coeffs)
+            .collect::<Vec<_>>();
+        let i_poly = linear_combination(&interpolated_polynomials[..], &etas).unwrap();
+
+        let i_comm = msm::<E>(&self.powers_of_g, &i_poly);
+
+        // Gathering commitments
+        let comm_vec = commitments.iter().map(|x| x.0).collect::<Vec<_>>();
+        let etas_repr = etas.iter().map(|e| e.into_bigint()).collect::<Vec<_>>();
+        let f_comm = VariableBase::msm(&comm_vec, &etas_repr);
+
+        let g2 = self.powers_of_g2[0];
+
+        if E::pairing(f_comm - i_comm.into_projective(), g2) == E::pairing(proof.0, zeros) {
+            Ok(())
+        } else {
+            Err(VerificationError)
+        }
+    }
+}
+
+fn interpolate_poly<E: PairingEngine>(
+    eval_points: &[E::Fr],
+    evals: &[E::Fr],
+    sca_inverse: &[E::Fr],
+    lang: &[DensePolynomial<E::Fr>],
+) -> DensePolynomial<E::Fr> {
+    let mut res = DensePolynomial::from_coefficients_vec(vec![E::Fr::zero()]);
+    for (j, (_x_j, y_j)) in eval_points.iter().zip(evals.iter()).enumerate() {
+        let l_poly = lang[j].mul(sca_inverse[j] * y_j);
+        res = (&res).add(&l_poly);
+    }
+    res
+}
+
+/// The polynomial in \\(\FF\\) that vanishes in all the points `points`.
```

Comment:
This takes quadratic time, no?

Reply:
Yes. It is quadratic to the number of evaluation points. In Gemini or Marlin, there is only a constant number of evaluation points.